Multiple unit tests failing under sanitizers due to memory leaks

# Reproducible on systems
1. System: Docker Ubuntu 16.04 Compiler: gcc version 8.4.0 
2. System: NixOS 20.09 Compiler: gcc version 7.5.0 
3. System: NixOS 20.09 Compiler: gcc version 8.4.0
4. System: NixOS 20.09 Compiler: gcc version 10.2.0

# Steps to reproduce
1. git clone https://github.com/actor-framework/actor-framework.git
2. cd actor-framework/
3. git checkout master
4. ./configure --sanitizers=address
5. cd build/
6. make
7. ctest

# Expected behavior
All tests pass with no issues reported

# Observed behavior
Ctest reports multiple test failures:
```
93% tests passed, 10 tests failed out of 144

Total Test time (real) =  10.83 sec

The following tests FAILED:
	  1 - actor_clock (Failed)
	  4 - actor_pool (Failed)
	  9 - aout (Failed)
	 32 - detail.group_tunnel (Failed)
	 35 - detail.local_group_module (Failed)
	 86 - mixin.sender (Failed)
	 89 - node_id (Failed)
	124 - uri (Failed)
	127 - custom_exception_handler (Failed)
	138 - io.remote_group (Failed)
```

When running for example:
ctest -VV -R "aout"

Output contains more descriptive:
```
: +----------------------------------------------------------------------+
9:                                 summary
9: +----------------------------------------------------------------------+
9: 
9:                         suites:  1
9:                         tests:   2
9:                         checks:  5
9:                         time:    15 ms
9:                         success: 100%
9: 
9: +----------------------------------------------------------------------+
9: ==10323==Processing thread 10300.
9: ==10323==Stack at 0x7ffeb1bc8000-0x7ffeb23c8000 (SP = 0x7ffeb23c5ae8).
9: ==10323==TLS at 0x7f9a2b814000-0x7f9a2b8150c0.
9: 
9: =================================================================
9: ==10300==ERROR: LeakSanitizer: detected memory leaks
9: 
9: Direct leak of 184 byte(s) in 1 object(s) allocated from:

```

# Thoughts
1. It's really critical to not leak any resources.
2. Maybe CI system is not configured properly to run test under sanitizers, because as I can see - this problem is not reported, or it leaks on this particular setup of os + compiler - but would be at least strange behavior.

# Additional problems
## I was not able to compile code under clang-8 - with output:
```
Reaping winning child 0x1038010 PID 12583 
Live child 0x1038010 (examples/simple_http_broker) PID 12585 
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_report_store4'
../libcaf_core/libcaf_core.so.0.18.0: undefined reference to `__asan_stack_free_7'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_register_globals'
../libcaf_core/libcaf_core.so.0.18.0: undefined reference to `__asan_stack_malloc_8'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_report_load_n'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_stack_malloc_0'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_memset'
../libcaf_core/libcaf_core.so.0.18.0: undefined reference to `__asan_stack_free_8'
../libcaf_core/libcaf_core.so.0.18.0: undefined reference to `__asan_report_store16'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_stack_malloc_1'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_report_load4'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_stack_malloc_6'
../libcaf_core/libcaf_core.so.0.18.0: undefined reference to `__asan_stack_malloc_7'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_report_store8'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_report_load2'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_unregister_globals'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_handle_no_return'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_init'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_stack_malloc_4'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_set_shadow_f5'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_memcpy'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_report_load1'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_alloca_poison'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_report_load16'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_report_store1'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_after_dynamic_init'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_stack_malloc_5'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_memmove'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_report_load8'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_report_store2'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_set_shadow_00'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_stack_free_5'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_stack_malloc_2'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_option_detect_stack_use_after_return'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_version_mismatch_check_v8'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_report_store_n'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_stack_free_6'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_before_dynamic_init'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_stack_malloc_3'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_allocas_unpoison'
../libcaf_io/libcaf_io.so.0.18.0: undefined reference to `__asan_set_shadow_f8'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

## I was not able to compile code under gcc with
./configure --sanitizers=address,undefined
or with
./configure --dev-mode
Report is:
```
 81%] Linking CXX executable caf-core-test
/usr/bin/ld: unrecognized option '--push-state--no-as-needed'
/usr/bin/ld: use the --help option for usage information
collect2: error: ld returned 1 exit status
```

I've found [this](https://stackoverflow.com/questions/50024731/ld-unrecognized-option-push-state-no-as-needed) about ld, but doesn't manage to fix by myself.

--- a/.ci/build.sh
+++ b/.ci/build.sh
@@ -5,6 +5,7 @@ set -e
 # This script expects the CAF source tree under 'sources', builds in 'build' and
 # installs to 'bundle'. Additionally, the script expects the CMake pre-load
 # script 'cmake-init.txt' in the current working directorys.
+BaseDir="$PWD"
 InitFile="$PWD/cmake-init.txt"
 SourceDir="$PWD/sources"
 BuildDir="$PWD/build"
@@ -23,6 +24,28 @@ else
   exit 1
 fi
 
+LeakSanCheck="
+#include <cstdlib>
+int main() {
+  int* ptr = new int(EXIT_SUCCESS);
+  return *ptr;
+}
+"
+
+# Check that LeakSanitizer works if configured for this build.
+if [ "${CAF_CHECK_LEAKS}" == "ON" ]; then
+  mkdir -p "$BaseDir/LeakCheck"
+  cd "$BaseDir/LeakCheck"
+  echo "${LeakSanCheck}" > check.cpp
+  c++ check.cpp -o check -fsanitize=address -fno-omit-frame-pointer
+  out=`./check 2>&1 | grep -o LeakSanitizer`
+  if [ "$out" != "LeakSanitizer" ]; then
+    echo "unable to detected memory leaks on this platform!"
+    return 1
+  fi
+  cd "$BaseDir"
+fi
+
 # Make sure all directories exist, then enter build directory.
 mkdir -p "$BuildDir"
 cd "$BuildDir"
@@ -34,4 +57,4 @@ if [ -z "$CAF_NUM_CORES" ]; then
 else
   $CMakeCommand --build . --target install -- -j $CAF_NUM_CORES
 fi
-$CTestCommand
+$CTestCommand --output-on-failure

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -40,6 +40,8 @@ is based on [Keep a Changelog](https://keepachangelog.com).
   dependency for `caf_core` to libatomic gets executables to compile and link as
   expected (#1153).
 - Fixed a regression for remote groups introduced in 0.18.0-rc.1 (#1157).
+- Fix memory leaks when deserializing URIs and when detaching the content of
+  messages (#1160).
 
 ## [0.18.0-rc.1] - 2020-09-09
 

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -31,13 +31,11 @@ config = [
             numCores: 4,
             tags: ['docker'],
             builds: ['debug', 'release'],
-            extraDebugFlags: ['CAF_SANITIZERS:STRING=address,undefined'],
         ]],
         ['centos-8', [
             numCores: 4,
             tags: ['docker'],
             builds: ['debug', 'release'],
-            extraDebugFlags: ['CAF_SANITIZERS:STRING=address,undefined'],
         ]],
         ['debian-9', [
             numCores: 4,
@@ -73,6 +71,7 @@ config = [
             numCores: 4,
             tags: ['docker'],
             builds: ['debug', 'release'],
+            extraDebugFlags: ['CAF_SANITIZERS:STRING=address'],
         ]],
         // One extra debug build with exceptions disabled.
         ['centos-7', [
@@ -119,6 +118,10 @@ config = [
     ],
     // Platform-specific environment settings.
     buildEnvironments: [
+        'fedora-32 && debug': [
+            'CAF_CHECK_LEAKS=ON',
+            'ASAN_OPTIONS=detect_leaks=1',
+        ],
         nop: [], // Dummy value for getting the proper types.
     ],
     // Default CMake flags by build type.

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -164,7 +164,7 @@ pipeline {
     environment {
         PrettyJobBaseName = env.JOB_BASE_NAME.replace('%2F', '/')
         PrettyJobName = "CAF/$PrettyJobBaseName #${env.BUILD_NUMBER}"
-        ASAN_OPTIONS = 'detect_leaks=0'
+        ASAN_OPTIONS = 'detect_leaks=1'
     }
     stages {
         stage('Checkout') {

--- a/libcaf_core/caf/typed_message_view.hpp
+++ b/libcaf_core/caf/typed_message_view.hpp
@@ -32,7 +32,7 @@ public:
     // nop
   }
 
-  explicit typed_message_view(message& msg) : ptr_(&msg.data()) {
+  explicit typed_message_view(message& msg) : ptr_(msg.ptr()) {
     // nop
   }
 

--- a/libcaf_core/caf/uri.hpp
+++ b/libcaf_core/caf/uri.hpp
@@ -30,6 +30,7 @@
 #include "caf/intrusive_cow_ptr.hpp"
 #include "caf/intrusive_ptr.hpp"
 #include "caf/ip_address.hpp"
+#include "caf/make_counted.hpp"
 #include "caf/string_view.hpp"
 #include "caf/variant.hpp"
 
@@ -273,8 +274,10 @@ struct inspector_access<uri> : inspector_access_base<uri> {
       return f.object(x).fields(f.field("value", get, set));
     } else {
       if constexpr (Inspector::is_loading)
-        if (!x.impl_->unique())
-          x.impl_.reset(new uri::impl_type);
+        if (!x.impl_->unique()) {
+          x.impl_.reset();
+          x.impl_ = make_counted<uri::impl_type>();
+        }
       return inspect(f, *x.impl_);
     }
   }
@@ -290,8 +293,10 @@ struct inspector_access<uri> : inspector_access_base<uri> {
       return detail::split_save_load(f, get, set);
     } else {
       if constexpr (Inspector::is_loading)
-        if (!x.impl_->unique())
-          x.impl_.reset(new uri::impl_type);
+        if (!x.impl_->unique()){
+          x.impl_.reset();
+          x.impl_ = make_counted<uri::impl_type>();
+        }
       return inspect(f, *x.impl_);
     }
   }

--- a/libcaf_core/caf/uri.hpp
+++ b/libcaf_core/caf/uri.hpp
@@ -274,10 +274,8 @@ struct inspector_access<uri> : inspector_access_base<uri> {
       return f.object(x).fields(f.field("value", get, set));
     } else {
       if constexpr (Inspector::is_loading)
-        if (!x.impl_->unique()) {
-          x.impl_.reset();
-          x.impl_ = make_counted<uri::impl_type>();
-        }
+        if (!x.impl_->unique())
+          x.impl_.reset(new uri::impl_type, false);
       return inspect(f, *x.impl_);
     }
   }
@@ -293,10 +291,8 @@ struct inspector_access<uri> : inspector_access_base<uri> {
       return detail::split_save_load(f, get, set);
     } else {
       if constexpr (Inspector::is_loading)
-        if (!x.impl_->unique()){
-          x.impl_.reset();
-          x.impl_ = make_counted<uri::impl_type>();
-        }
+        if (!x.impl_->unique())
+          x.impl_.reset(new uri::impl_type, false);
       return inspect(f, *x.impl_);
     }
   }

--- a/libcaf_core/src/detail/group_tunnel.cpp
+++ b/libcaf_core/src/detail/group_tunnel.cpp
@@ -120,15 +120,17 @@ void group_tunnel::stop() {
   CAF_LOG_TRACE("");
   CAF_LOG_DEBUG("stop group tunnel:" << CAF_ARG2("module", module().name())
                                      << CAF_ARG2("identifier", identifier_));
-  auto hdl = actor{};
+  auto worker_hdl = actor{};
+  auto intermediary_hdl = actor{};
   auto subs = subscriber_set{};
   auto cache = cached_message_list{};
-  auto stopped = critical_section([this, &hdl, &subs, &cache] {
+  auto stopped = critical_section([&, this] {
     using std::swap;
     if (!stopped_) {
       stopped_ = true;
       swap(subs, subscribers_);
-      swap(hdl, worker_);
+      swap(worker_hdl, worker_);
+      swap(intermediary_hdl, intermediary_);
       swap(cache, cached_messages_);
       return true;
     } else {
@@ -136,7 +138,7 @@ void group_tunnel::stop() {
     }
   });
   if (stopped) {
-    anon_send_exit(hdl, exit_reason::user_shutdown);
+    anon_send_exit(worker_hdl, exit_reason::user_shutdown);
     if (!subs.empty()) {
       auto bye = make_message(group_down_msg{group{this}});
       for (auto& sub : subs)

--- a/libcaf_core/src/detail/local_group_module.cpp
+++ b/libcaf_core/src/detail/local_group_module.cpp
@@ -163,7 +163,7 @@ expected<group> local_group_module::get(const std::string& group_name) {
     auto ptr = make_counted<impl>(this, group_name);
     ptr->intermediary_ = system().spawn<intermediary_actor, hidden>(ptr);
     instances_.emplace(group_name, ptr);
-    return group{ptr.release()};
+    return group{std::move(ptr)};
   }
 }
 

--- a/libcaf_core/src/detail/message_data.cpp
+++ b/libcaf_core/src/detail/message_data.cpp
@@ -69,6 +69,7 @@ message_data* message_data::copy() const {
     auto& meta = gmos[id];
     // TODO: exception handling.
     meta.copy_construct(dst, src);
+    ++ptr->constructed_elements_;
     src += meta.padded_size;
     dst += meta.padded_size;
   }

--- a/libcaf_core/test/detail/group_tunnel.cpp
+++ b/libcaf_core/test/detail/group_tunnel.cpp
@@ -83,8 +83,9 @@ struct testee_state {
 
 behavior testee_impl(stateful_actor<testee_state>* self) {
   return {
-    [=](put_atom, int x) { self->state.x = x; },
-    [=](get_atom) { return self->state.x; },
+    [self](put_atom, int x) { self->state.x = x; },
+    [self](get_atom) { return self->state.x; },
+    [self](const group_down_msg&) { self->quit(); },
   };
 }
 
@@ -103,9 +104,11 @@ struct fixture : test_coordinator_fixture<> {
     // Groups keep their subscribers alive (on purpose). Since we don't want to
     // manually kill all our testee actors, we simply force the group modules to
     // stop here.
+    uut->stop();
     for (auto& kvp : uut->instances)
       kvp.second->stop();
     sys.groups().get_module("local")->stop();
+    run();
   }
 
   void make_unconnected() {

--- a/libcaf_core/test/message.cpp
+++ b/libcaf_core/test/message.cpp
@@ -61,6 +61,24 @@ CAF_TEST(messages allow index - based access) {
   CAF_CHECK_EQUAL(msg.cdata().get_reference_count(), 1u);
 }
 
+CAF_TEST(message detach their content on mutating access) {
+  CAF_MESSAGE("Given to messages pointing to the same content.");
+  auto msg1 = make_message("one", uint32_t{1});
+  auto msg2 = msg1;
+  CAF_CHECK_EQUAL(msg1.cdata().get_reference_count(), 2u);
+  CAF_CHECK_EQUAL(msg1.cptr(), msg2.cptr());
+  CAF_MESSAGE("When calling a non-const member function of message.");
+  msg1.ptr();
+  CAF_MESSAGE("Then the messages point to separate contents but remain equal.");
+  CAF_CHECK_NOT_EQUAL(msg1.cptr(), msg2.cptr());
+  CAF_CHECK_EQUAL(msg1.cdata().get_reference_count(), 1u);
+  CAF_CHECK_EQUAL(msg2.cdata().get_reference_count(), 1u);
+  CAF_CHECK(msg1.match_elements<std::string, uint32_t>());
+  CAF_CHECK(msg2.match_elements<std::string, uint32_t>());
+  CAF_CHECK_EQUAL(msg1.get_as<std::string>(0), msg2.get_as<std::string>(0));
+  CAF_CHECK_EQUAL(msg1.get_as<uint32_t>(1), msg2.get_as<uint32_t>(1));
+}
+
 CAF_TEST(compare_custom_types) {
   s2 tmp;
   tmp.value[0][1] = 100;

